modules/src/events.rs

```diff
@@ -0,0 +1,250 @@
+    OpenConfirmConnection(ConnectionEvents::OpenConfirm),
+
+    OpenInitChannel(ChannelEvents::OpenInit),
+    OpenTryChannel(ChannelEvents::OpenTry),
+    OpenAckChannel(ChannelEvents::OpenAck),
+    OpenConfirmChannel(ChannelEvents::OpenConfirm),
+    CloseInitChannel(ChannelEvents::CloseInit),
+    CloseConfirmChannel(ChannelEvents::CloseConfirm),
+
+    SendPacketChannel(ChannelEvents::SendPacket),
+    ReceivePacketChannel(ChannelEvents::ReceivePacket),
+    AcknowledgePacketChannel(ChannelEvents::AcknowledgePacket),
+    CleanupPacketChannel(ChannelEvents::CleanupPacket),
+    TimeoutPacketChannel(ChannelEvents::TimeoutPacket),
+
+    TimeoutTransfer(TransferEvents::Timeout),
+    PacketTransfer(TransferEvents::Packet),
+    ChannelClosedTransfer(TransferEvents::ChannelClosed),
+}
+
+impl IBCEvent {
+    pub fn to_json(&self) -> String {
+        serde_json::to_string(self).unwrap()
+    }
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct RawObject {
+    pub height: block::Height,
+    pub action: String,
+    pub idx: usize,
+    pub events: HashMap<String, Vec<String>>,
+}
+
+impl RawObject {
+    pub fn new(
+        height: block::Height,
+        action: String,
+        idx: usize,
+        events: HashMap<String, Vec<String>>,
+    ) -> RawObject {
+        RawObject {
+            height,
+            action,
+            idx,
+            events,
+        }
+    }
+}
+
+pub fn extract_events<S: ::std::hash::BuildHasher>(
+    events: &HashMap<String, Vec<String>, S>,
+    action_string: &str,
+) -> Result<(), BoxError> {
+    if let Some(message_action) = events.get("message.action") {
+        if message_action.contains(&action_string.to_owned()) {
+            return Ok(());
+        }
+        return Err("Missing action string".into());
+    }
+    Err("Incorrect Event Type".into())
+}
+
+//Takes events in the form
+//  "events":{
+//      "connection_open_init.client_id":["testclient","testclientsec"],
+//      "connection_open_init.connection_id":["ancaconnonetest","ancaconnonetestsec"],
+//      "connection_open_init.counterparty_client_id":["testclientsec","testclientsecsec"],
+//      "create_client.client_id":["testclientthird"],
+//      "create_client.client_type":["tendermint"],
+//      "message.action":["connection_open_init","create_client","connection_open_init"],
+//      "message.module":["ibc_connection","ibc_client",ibc_connection"],
+//      "message.sender":["cosmos187xxg4yfkypl05cqylucezpjvycj24nurvm8p9","cosmos187xxg4yfkypl05cqylucezpjvycj24nurvm8p9","cosmos187xxg4yfkypl05cqylucezpjvycj24nurvm8p9","cosmos187xxg4yfkypl05cqylucezpjvycj24nurvm8p9"],"tm.event":["Tx"],"transfer.amount":["5000stake"],"transfer.recipient":["cosmos17xpfvakm2amg962yls6f84z3kell8c5lserqta"],"tx.hash":["A9E18AE3909F22232F8DBDB1C48F2FECB260A308A2D157E8832E901D45950605"],
+//      "tx.height":["35"]
+// and returns:
+// [{"connection_open_init", 0}, {"create_client", 0}, {"connection_open_init", 1}]
```

Comment:
What is the reasoning behind this format of return value?

Reply:
To help extract events in the order of their occurrence. The `message.action` is a vector of events e.g. 
`message.action = [x, y,  x]` 
where you have two events of type `x`. 
For each event type, each attribute is an array, eg: 
```
x.attr1 = ["a", "b"]
x.attr2 = ["A", "B"]
y.attr1 = ["c"]
```
 So if you would want to get all events of type `x` you iterate over the attribute arrays `x.attr1` and `x.attr2`
But if you want them in order then you first extract x then y then x. The helper tells you at what index in the x and y vectors you should take the values from.

I was thinking to merge this with `build_event()` as it can all be done in one pass.